File: app/Ligare/Pretty.hs
module Ligare.Pretty where

import Ligare.Core.Syntax

pretty :: Term -> String
pretty (Var i) = "$" ++ show i
pretty (Lam body) = "λ. " ++ pretty body
pretty (App f a) = "(" ++ pretty f ++ " " ++ pretty a ++ ")"
pretty (LitInt n) = show n
pretty (Universe UData) = "data"
pretty (Universe UProp) = "prop"
pretty (Universe UTheorem) = "theorem"
pretty (Universe UProof) = "proof"
pretty (Pi "" a b) = "(" ++ pretty a ++ " -> " ++ pretty b ++ ")"
pretty (Pi name a b) = "(Pi " ++ name ++ " : " ++ pretty a ++ " => " ++ pretty b ++ ")"
pretty (Builtin s) = s
pretty (PrimOp Add) = "+"
pretty (PrimOp Sub) = "-"
pretty (PrimOp Mul) = "*"
pretty (PrimOp Div) = "/"
pretty (PrimOp Mod) = "%"
pretty (PrimOp Eq) = "=="
pretty (PrimOp Lt) = "<"
pretty (PrimOp Gt) = ">"
pretty (PrimOp Le) = "<="
pretty (PrimOp Ge) = ">="
pretty (PrimOp Neq) = "/="
pretty (LitBool b) = show b
pretty (Let name val body mconstr) =
  let constrStr = case mconstr of
        Just c -> " : " ++ pretty c
        Nothing -> ""
   in "let " ++ name ++ constrStr ++ " = " ++ pretty val ++ " in " ++ pretty body
pretty (IfThenElse cond tbranch fbranch) =
  "if " ++ pretty cond ++ " then " ++ pretty tbranch ++ " else " ++ pretty fbranch
pretty (Refine name parent p) =
  "constraint " ++ name ++ " = " ++ pretty parent ++ " (x => " ++ pretty p ++ ")"
pretty (Annot t c) = "(" ++ pretty t ++ " : " ++ pretty c ++ ")"
pretty (ByProof t proof) = "(" ++ pretty t ++ " by " ++ pretty proof ++ ")"
pretty AutoProof = "auto"
pretty RefParam = "x"
pretty (ProofBlock t) = "proof { " ++ pretty t ++ " }"
pretty (Func name params mRet preconds postconds body) =
  "func "
    ++ name
    ++ "("
    ++ prettyParams params
    ++ ")"
    ++ maybe "" (\r -> " : " ++ pretty r) mRet
    ++ concat [" pre: " ++ pretty p | p <- preconds]
    ++ concat [" post: " ++ pretty p | p <- postconds]
    ++ " = "
    ++ pretty body
  where
    prettyParams [] = ""
    prettyParams [(n, Nothing)] = n
    prettyParams [(n, Just c)] = n ++ " : " ++ pretty c
    prettyParams ((n, mc) : rest) =
      (case mc of Nothing -> n; Just c -> n ++ " : " ++ pretty c)
        ++ ", "
        ++ prettyParams rest

File: tests/pretty.rs
mod common;

use common::{bin, leak_bump, s};
use ligare::core::pool::TermArena;
use ligare::core::syntax::{PrimOp, Term};
use ligare::pretty::pretty;

fn a() -> (&'static bumpalo::Bump, TermArena<'static>) {
    let b = leak_bump();
    (b, TermArena::new(b))
}

#[test]
fn integer() {
    assert_eq!(pretty(&Term::LitInt(42)), "42");
}

#[test]
fn lambda() {
    assert_eq!(pretty(&Term::Lam(&Term::Var(0))), "λ. $0");
}

#[test]
fn if_() {
    let (_bump, arena) = a();
    let t = arena.if_then_else(arena.lit_bool(true), arena.lit_int(1), arena.lit_int(0));
    assert_eq!(pretty(t), "if true then 1 else 0");
}

#[test]
fn let_() {
    let (_bump, arena) = a();
    let t = arena.let_(s(&arena, "x"), arena.lit_int(5), arena.var(0), None);
    assert_eq!(pretty(t), "let x = 5 in $0");
}

#[test]
fn annot() {
    let (_bump, arena) = a();
    let t = arena.annot(arena.lit_int(5), arena.builtin(s(&arena, "int")));
    assert_eq!(pretty(t), "(5 : int)");
}

#[test]
fn refine() {
    let (_bump, arena) = a();
    let pred = bin(&arena, PrimOp::Ge, arena.ref_param(), arena.lit_int(0));
    let t = arena.refine(s(&arena, ""), arena.builtin(s(&arena, "int")), pred);
    assert_eq!(pretty(t), "int where (x => ((>= x) 0))");
}
